### ISSUE TITLE
Make error scope objects have same shape as DOMExceptions?

### ISSUE THREAD

**Issue by juj** [gpuweb/gpuweb#1884]
Marshalling `device.popErrorScope()` in to Wasm is a bit tedious, given the polymorphic nature of the error object that is resolved in the handler.

Also if the Promise returned by `device.popErrorScope()` catches instead of resolving, in Chrome at least the error scope catch handler will be called with a DOMException error, which has the fields `.code`, `.message` and `.name`. This occurs on the "connection to GPU lost" scenario.

Would it be possible to make the `GPUOutOfMemoryError` and `GPUValidationError` objects (and all future error objects) both have the same shape as a DOMException? If they all had code/message/name fields, it would be straightforward to marshal an error over to wasm and/or log an error to console, as the errors would always come with standard fields to them.

This way JS->Wasm marshalling code would become future compatible, say if a new error type was introduced, then existing bindings code would still be able to could do that without needing an update. 

In the current form one ends up doing something like `if (error instanceof GPUOutOfMemoryError) {} else if (error instanceof GPUValidationError) {} else if (error instanceof DOMException) {} else { console.error(error.toString);` type of handling to reason about the errors, which is a bit awkward, and if a new error is introduced, one can only manage it as a .toString()ed thing.

**Comment by juj** [comment]
Related - not sure if this is a spec bug or an implementation bug, but I find having to write

```
try {
  device.popErrorScope().then(handleOomError).catch(handleOomError);
} catch(e) {
  handleOomError(e);
}
```

in this test to catch OOM errors without leaking any through. Looks like Chrome can either resolve, reject or throw a call to popErrorScope().

See https://bugs.chromium.org/p/chromium/issues/detail?id=1224835#c4

**Comment by kainino0x** [comment]
> in this test to catch OOM errors without leaking any through. Looks like Chrome can either resolve, reject or throw a call to popErrorScope().

The spec here isn't complete yet, but this is [in the spec](https://gpuweb.github.io/gpuweb/#dom-gpudevice-poperrorscope). A rejection/exception *does not* indicate a (recoverable) OOM, it only indicates either (1) the device is lost, or (2) the error scope stack was empty. This _should_ be what Chrome implements. Recoverable OOM is always a return value of GPUOutOfMemoryError.

The entry point (and all IDL entry points that return `Promise<>`) should also never throw exceptions, it should only return rejected promises. If Chrome doesn't behave that way it's a bug.

**Comment by juj** [comment]
That does seem to be the case, it was just the Chrome DevTools that was triggering unexpectedly.

The question about unifying the error type shapes remains.

There is also the [GPUDeviceLostInfo](https://www.w3.org/TR/webgpu/#errors-and-debugging) object, which is very similar to the `GPUOutOfMemoryError` and `GPUValidationError` objects. It would be nice if they each had the same shape.

**Comment by kainino0x** [comment]
@juj Now that we've clarified the API here, would this change still be helpful? Since getting a DOMException and getting a GPUOutOfMemoryError indicate very different things, I don't think it makes sense to have any kind of common handler (like your `handleOomError` example). So I think the only reason to make them similar would be that it would _happen_ to allow sharing of some JS->C/++ marshalling code and maybe some C/++ interface.

**Comment by Kangz** [comment]
@juj any feedback on what @kainino0x said? Otherwise I think we should close this issue.

**Comment by juj** [comment]
Yes, there is still a difficulty with marshalling error objects to wasm. In current form, there is no good way to write code that is future proof - in case the api introduces new errors to browser implementations, all webgpu api utilizing code needs to be updated & recompiled with an updated Emscripten compiler/wasm->js binding layer to add support for that error code to the application.

For example, currently @kainino0x 's binding layer does this:

https://github.com/emscripten-core/emscripten/blob/ad6e356df588646eb2569c8e9100b8991db799e8/src/library_webgpu.js#L568-L578

If a new error type is introduced, that code will output "error type 0" to calling wasm code for any such new errors.

My bindings layer does this:

https://github.com/juj/wasm_webgpu/blob/dc00d05fd678f19ae36b344ba2b13a38d85f40bd/lib/lib_webgpu.js#L222-L227

so it reports `WGPU_ERROR_FILTER_OUT_OF_MEMORY (185) or `WGPU_ERROR_FILTER_VALIDATION (186)` for the two currently recognized errors, and `-1` for any unknown future errors.

Both layers showcase the same drawback that they need to `if()` branch through all possible error types.

Maybe in current form one could run `error.toString()` on the error object, to get a string like `[object GPUOutOfMemoryError]`, and then hash the string to get an integer code, and then marshal that code over to wasm. But my understanding is that different browsers may return a different string with `error.toString()`, so the code should do a browser-specific path for each object.

If `GPUOutOfMemoryError` and `GPUValidationError` and all other errors would have the same code/message/name fields that a `DOMException` has, then a common marshalling code could be used for all these types, that would be future proof so that all webgpu user code would not need recompiling with an updated compiler if the spec&implementation changes in the future with new error types.

**Comment by Kangz** [comment]
@juj, in the current shape of webgpu.h I think you can already solve the issue you have without a change to the WebGPU API.

 - `WGPUErrorType` has `WGPUErrorType_Unknown` which the C++-JS bindings can use when the browser returns something it doesn't know. If a new type of error is added to WebGPU, applications will need special code if they want to do specific handling for them, so they need to be recompiled and can update/patch emscripten at that point.
 - It is extremely likely (if not guaranteed) that any new error type with a message that is added, adds the message as a `message` member. So you can do `message = error['message'] ?? ''` in the unknown error type case.
 - Linearly scanning through error types is a bit sad, but at the same time we don't expect many if any new error types to be added, so w/e?

That's why I think we should lean towards doing nothing, so that's one less thing to do for v1.

**Comment by juj** [comment]
> It is extremely likely (if not guaranteed) that any new error type with a message that is added, adds the message as a message member.

Why would only future errors get a message field? Why not put it on all existing fields? Seems like a simple and noncontroversial thing to do.

> If a new type of error is added to WebGPU, applications will need special code if they want to do specific handling for them, so they need to be recompiled and can update/patch emscripten at that point.

The point of this issue was to specifically ask for a mechanism that would not need a new Emscripten update to be able to handle a new error; and would not need code to even be recompiled if the application does nothing else with the error except than to print it out to the user.

Seems like both of those challenges can be solved by adding standardized error code and message fields, rather than rely on the type of the error object to pass errors.

**Comment by juj** [comment]
It is good to note that there is a major distinction between

- "I (as game developer) can recompile my code to add support for handling a new WebGPU error type to it."

vs

- "Emscripten authors first have to add support for the new WebGPU error type, after which I (as game developer) will first need to update to that patched Emscripten version, and only then I can recompile my code to add support for handling a new error event to it."

Moreover, neither of the above would be needed for those classes of errors where the user's handling mechanism would constitute of a "print the error out to the web console and either abort/keep on trucking" - those could be handled generally with no code rebuilds whatsoever. That would be quite more debuggable on live sites compared to sites printing "unknown error" messages in the future.

> So you can do message = error['message'] ?? '' in the unknown error type case.

It is not particularly hygienic for (Emscripten) developers to try to guess future API IDL in advance and try to predict what fields will be likely to appear in an upcoming spec. Though that is already what I am doing and what Emscripten bindings also seem to be doing (though looks like accidentally so, since Emscripten bindings assumes everything not a GPUOutOfMemoryError would be a GPUValidationError that would have such a message field).

But I do appreciate being up front about shipping fatigue coming on to get v1 out the door, so if "sad, but w/e" is the best that can happen here, it is no biggie. This is by far not on our critical path, but just an observation on how error handling could be future compatible so that Emscripten developers will not need to interfere updating Emscripten WebGPU versions to support users.

---

@kainino0x @shrekshao :

Current Emscripten WebGPU bindings will in one site crash in development builds if future browsers will produce a new error type: https://github.com/emscripten-core/emscripten/blob/500d8267b4eb22fcc6606aa695a1a076c0f40375/src/library_webgpu.js#L736 . They only ever expect seeing `GPUOutOfMemoryError` or `GPUValidationError` errors to be produced.

On another code site it will treat a new error type as being as if no error occurred: https://github.com/emscripten-core/emscripten/blob/500d8267b4eb22fcc6606aa695a1a076c0f40375/src/library_webgpu.js#L804 . (e.g. [0 == NoError](https://github.com/emscripten-core/emscripten/blob/500d8267b4eb22fcc6606aa695a1a076c0f40375/src/library_webgpu.js#L106)) 

On a third code site @kainino0x has added a TODO to figure out how to extend the type checking: https://github.com/emscripten-core/emscripten/blob/500d8267b4eb22fcc6606aa695a1a076c0f40375/src/library_webgpu.js#L1780

It may be good to resolve those at some point to at least prevent Emscripten WebGPU page builds from today misbehaving on WebGPU browser implementations from the future.

---

In my wasm_webgpu bindings unknown future errors will get an error code -1 and if there is a message field, that will be printed:

https://github.com/juj/wasm_webgpu/blob/0051c71c13fc22f59ef02accde934f81ad0dba76/lib/lib_webgpu.js#L343-L352

Also users cannot push an error scope of a future error on their own without first obtaining an updated bindings library version:

https://github.com/juj/wasm_webgpu/blob/0051c71c13fc22f59ef02accde934f81ad0dba76/lib/lib_webgpu.js#L338

**Comment by kainino0x** [comment]
It does seem worthwhile to get rid of the type hierarchy and replace it with an enum. Adding a numeric code to the JS API seems nice for WASM, but perhaps caters too much to it. Without it, WASM bindings can just pass in a raw string for unknown enum values. (The WASM-heap allocation of that string can be scoped to a callback just like the `message` string.)

**Comment by kainino0x** [comment]
(marking as pseudo-editorial for burndown when I get back to spec work)

**Comment by kainino0x** [comment]
Resolved in #2853
